Wangscape/AlphaCalculator.cpp

```diff
@@ -0,0 +1,85 @@
+#include "AlphaCalculator.h"
+#include <algorithm>
+#include <numeric>
+
+
+
+AlphaCalculator::AlphaCalculator()
+{
+}
+
+
+AlphaCalculator::~AlphaCalculator()
+{
+}
+
+void AlphaCalculator::calculatePixelAlphaLinear(WeightVector& weights, AlphaVector& alphas)
+{
+    int err = 255;
+    float total_weight = 0.f;
+    for (auto& weight : weights)
+    {
+        weight = std::max(0.f, weight);
+        total_weight += weight;
+    }
+    if (total_weight <= 0.f)
+    {
+        // This shouldn't happen, but it is easy to make weight masks
+        // where this does happen.
+        for (auto& alpha : alphas)
+        {
+            alpha = 0;
+        }
+        return;
+    }
+    for (size_t i = 0; i < weights.size(); i++)
+    {
+        int alpha = (int)((255 * weights[i]) / total_weight);
```

Comment:
I agree there should be more assertions made, but I don't think this is a suitable site for them. Most of the calculations performed by Wangscape will eventually become user-configurable, and when they go wrong, it will be important to show the context of the incorrect calculations to the user. The best way to do that won't be to terminate with an exception, but to output debug tilesets showing the important steps of the calculation. Writing a message in the log file would also be a good idea.